sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -58,6 +58,108 @@
+        del self._cache[key]
+        self.merge_intervals_by_name_version()
+
+    def __iter__(self) -> t.Iterator[SnapshotId]:
+        return iter(self._cache)
+
+    def __len__(self) -> int:
+        return len(self._cache)
+
+    @property
+    def snapshots(self) -> t.List[Snapshot]:
+        return list(self.values())
+
+    @property
+    def is_hydrated(self) -> bool:
+        return all(snapshot.seed_is_hydrated for snapshot in self.values())
+
+    @property
+    def snapshot_intervals(self) -> t.List[SnapshotIntervals]:
+        return [snapshot.snapshot_intervals for snapshot in self.values()]
+
+    def get_cached_snapshots(self, snapshots: t.List[Snapshot]) -> t.List[Snapshot]:
+        cached_snapshots = []
+        for snapshot in snapshots:
+            self.add_snapshot(snapshot)
+            cached_snapshots.append(self[snapshot.snapshot_id])
+        return cached_snapshots
+
+    def add_snapshot(self, snapshot: Snapshot) -> None:
+        self[snapshot.snapshot_id] = snapshot
+
+    def remove_interval(
+        self, all_snapshots: t.Iterable[Snapshot], start: TimeLike, end: TimeLike
+    ) -> None:
+        for snapshot in all_snapshots:
+            self[snapshot.snapshot_id].remove_interval(start, end)
+
+    def get_snapshots_merged_intervals_by_id(
+        self, intervals: t.Optional[t.Iterable[SnapshotIntervals]] = None
+    ) -> t.Dict[SnapshotId, Snapshot]:
+        intervals_by_snapshot_id = {x.snapshot_id: x for x in intervals or self.snapshot_intervals}
+
+        result = {}
+        for snapshot in self.snapshots:
+            # We want to make sure that intervals for snapshots are not shared when doing the merge intervals
+            # so we do a deep copy of the snapshot
+            snapshot = snapshot.copy(deep=True, update={"intervals_": [], "dev_intervals_": []})
+            if snapshot.snapshot_id in intervals_by_snapshot_id:
+                snapshot.merge_intervals(intervals_by_snapshot_id[snapshot.snapshot_id])
+            result[snapshot.snapshot_id] = snapshot
+        return result
+
+    def contains(self, keys: t.Iterable[SnapshotIdLike]) -> bool:
+        return set(k.snapshot_id for k in keys).issubset(self)
+
+    def to_dict(self, keys: t.Optional[t.Iterable[SnapshotIdLike]]) -> t.Dict[SnapshotId, Snapshot]:
+        return (
+            {k: v for k, v in self.items() if k in {key.snapshot_id for key in keys}}
+            if keys is not None
+            else dict(self)
+        )
+
+    def merge_intervals_by_name_version(
+        self, intervals: t.Optional[t.Iterable[SnapshotIntervals]] = None
+    ) -> None:
+        interval_mapping: t.Dict[t.Tuple[str, str], t.List[SnapshotIntervals]] = defaultdict(list)
+        intervals = intervals or self.snapshot_intervals
+        for interval in intervals:
+            interval_mapping[(interval.name, interval.version)].append(interval)
+        for snapshot in self.snapshots:
+            snapshot.init_intervals()
```

Comment:
Should the init happen lazily as part of `merge_intervals`? Not sure why do we need a dedicated method for this

Reply:
Removed since intervals are now empty list by default: https://github.com/TobikoData/sqlmesh/pull/1018/commits/3323361b95251e980a2b3fea7244a3c23840e779